Order 기본 api 개발

axhtl opened the pull request:
order 생성, 조회 기본 api 개발

axhtl committed:
feat: 주문생성 및 조회 기본 로직 개발

주문 생성 및 조회 기능에 필요한 기능 로직을 구현하였습니다.
현재는 DB에 주문을 저장하고 조회할 수 있지만, 추가적인 기능(예: 주문 시각 자동 추가)은 아직 구현되지 않았습니다. 향후에 이런 추가 기능을 구현할 예정입니다.

axhtl committed:
refactor: 컨트롤러, 서비스, 레포지토리 반환값을 DTO로 변경

axhtl committed:
refactor: 컨트롤러, 서비스 반환값을 DTO로 변경

dev-mentoring commented on src/main/java/org/store/clothstar/order/controller/OrderController.java:
```diff
@@ -0,0 +1,30 @@
+	private final OrderService orderService;
+
+	@PostMapping("/v1/orders")
+	public CreateOrderRequest saveOrder(CreateOrderRequest createOrderRequest) {
```
값을 저장하는 로직 -> Post

- @ RequestBody 로 값을 받기  
```
@RequestBody CreateOderRequest
```

@RequestParam
@Pathvariable

수빈님께서 찾아보셔야 하는 내용은 위와같이 통신 방법에서 값을 전달받을 수 있는 방법들에 대해서 학습해보시고 차이점을 찾아보시면 좋습니다!

dev-mentoring replied:
@Valid, @Valided 가 있는데 내용은 노션에 작성해둘게요!

dev-mentoring commented on src/main/java/org/store/clothstar/order/dto/CreateOrderRequest.java:
```diff
@@ -0,0 +1,41 @@
+import lombok.Setter;
+
+@Getter
+@Setter
```
@RequestBody를 사용하게되면 @Setter를 사용하지 않으셔도 됩니다!

dev-mentoring commented on src/main/java/org/store/clothstar/order/controller/OrderController.java:
```diff
@@ -0,0 +1,30 @@
+
+	@PostMapping("/v1/orders")
+	public CreateOrderRequest saveOrder(CreateOrderRequest createOrderRequest) {
+		return orderService.saveOrder(createOrderRequest);
```
@Setter를 사용하게 될 경우,
createOrderRequest.shipingAmt = 10;
값을 변경이 가능합니다.

dev-mentoring commented on src/main/java/org/store/clothstar/order/domain/Order.java:
```diff
@@ -0,0 +1,40 @@
+	private Long memberId;
+	private Long deliveryId;
+	private String createdDt;
+	@DateTimeFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss")
```
이 어노테이션은 컨트롤러의 DTO에서 값을 포맷팅해서 가져오면 더 좋습니다!

dev-mentoring commented on src/main/java/org/store/clothstar/order/dto/OrderResponse.java:
```diff
@@ -0,0 +1,27 @@
+
+@AllArgsConstructor
+@Getter
+@Setter
```
Setter가 어디에 들어가고 안들어갈지 전략을 다시 세우시면 좋겠습니다!

dev-mentoring commented on src/main/java/org/store/clothstar/order/service/OrderService.java:
```diff
@@ -0,0 +1,30 @@
+			order.getShippingAmt(),
+			order.getProductsAmt(), order.getPaymentMethod(), order.getPaymentAmt());
+		return orderResponse;
+	}
```
각 메서드 나열방식에서 Get, Post, Put, Patch, Delete와 같이 순서가 일정하게 반영되면 좋습니다

dev-mentoring commented on src/main/java/org/store/clothstar/order/service/OrderService.java:
```diff
@@ -0,0 +1,30 @@
+	}
+
+	public CreateOrderRequest saveOrder(CreateOrderRequest createOrderRequest) {
+		orderRepository.saveOrder(createOrderRequest);
```
CreateOrderRequest는 DTO -> 데이터를 옮기는 역할을 합니다.
값을 저장하고 있는 역할까지도 하고 있습니다.

CreateOrderRequestDTO가 Order 객체로 변환한 상태에서 Repository에 저장이 되어야 합니다.

dev-mentoring replied:
Order 객체 name 필드가 있다고 가정해보겠습니다.

OrderRequest에도 name 필드가 있습니다.

Order가 저장이 될때는 OrderRequest.name + 현재시간의 양식으로 들어가야합니다.

이럴때 저희는 외부에서 받아온 OrderRequest DTO를 가공하여 Order를 저장해야합니다.

dev-mentoring commented on src/main/java/org/store/clothstar/order/service/OrderService.java:
```diff
@@ -0,0 +1,30 @@
+			order.getDeliveryId(), order.getCreatedDt(), order.getCreatedAt(), order.getStatus(),
+			order.getShippingAmt(),
+			order.getProductsAmt(), order.getPaymentMethod(), order.getPaymentAmt());
+		return orderResponse;
```
Order를 그대로 던지지 않고 OrderResponse의 객체로 던지는 부분은 잘하셨습니다.

OrderResponse는 여기서 생성을 하면 안됩니다.

Order의 메서드로 만들어 줘야합니다.

dev-mentoring commented on src/main/java/org/store/clothstar/order/domain/Order.java:
```diff
@@ -0,0 +1,40 @@
+		this.paymentMethod = paymentMethod;
+		this.paymentAmt = paymentAmt;
+	}
+}
```
```
public OrderResponse toResponseDTO(Long orderId, Long memberId, Long deliveryId, String createdDt,LocalDateTime createdAt, String status,int shippingAmt, int productsAmt, PaymentMethod paymentMethod, int paymentAmt) {

    return new OrderResposne (
...
)
```

dev-mentoring commented on src/main/java/org/store/clothstar/order/service/OrderService.java:
```diff
@@ -0,0 +1,30 @@
+	public OrderResponse getOrder(Long orderId) {
+		Order order = orderRepository.getOrder(orderId);
+		OrderResponse orderResponse = new OrderResponse(order.getOrderId(), order.getMemberId(),
+			order.getDeliveryId(), order.getCreatedDt(), order.getCreatedAt(), order.getStatus(),
```
축약어 사용하지마시고 다 작성부탁드립니다!

dev-mentoring commented on src/main/resources/mappers/Order.xml:
```diff
@@ -0,0 +1,36 @@
+        #{productsAmt}, #{paymentMethod}, #{paymentAmt});
+    </insert>
+
+    <!--    <resultMap type="org.store.clothstar.order.domain.Order" id="order">-->
+    <!--        <result property="orderId" column="order_id"/>-->
+    <!--        <result property="memberId" column="member_id"/>-->
+    <!--        <result property="deliveryId" column="delivery_id"/>-->
+    <!--        <result property="createdDt" column="created_dt"/>-->
+    <!--        <result property="createdAt" column="created_at"/>-->
+    <!--        <result property="status" column="status"/>-->
+    <!--        <result property="shippingAmt" column="shipping_amt"/>-->
+    <!--        <result property="productsAmt" column="products_amt"/>-->
+    <!--        <result property="paymentMethod" column="payment_method"/>-->
+    <!--        <result property="paymentAmt" column="payment_amt"/>-->
+    <!--    </resultMap>-->
+
+    <!--    <select id="getOrder" resultMap="order">-->
+    <!--        select * from orderline-->
+    <!--        where order_id = #{orderId}-->
+    <!--    </select>-->
```
필요없는 코드는 삭제부탁드립니다!

dev-mentoring commented on src/main/java/org/store/clothstar/order/service/OrderService.java:
```diff
@@ -0,0 +1,30 @@
+	public OrderResponse getOrder(Long orderId) {
+		Order order = orderRepository.getOrder(orderId);
+		OrderResponse orderResponse = new OrderResponse(order.getOrderId(), order.getMemberId(),
+			order.getDeliveryId(), order.getCreatedDt(), order.getCreatedAt(), order.getStatus(),
+			order.getShippingAmt(),
+			order.getProductsAmt(), order.getPaymentMethod(), order.getPaymentAmt());
```
new OrderResponse (
  orderId = order.getOrderId(),
  meberId = order.getMemberId(),
// memberId는 OrderResponse의 필드 이름으로 가정합니다.

dev-mentoring commented on src/main/resources/sql/order.sql:
```diff
@@ -0,0 +1,29 @@
+FROM orderline;
+
+
+INSERT INTO orderline (order_id, member_id, delivery_id, created_dt, created_at, status, shipping_amt, products_amt,
+                       payment_method, payment_amt)
+VALUES ('1', '2', '3', 'a', now(), 'b', '4', '5', '6', '7');
```
테스트 데이터는 실 데이터와 최대한 동일하게 값을 넣어주셔야 좋습니다!

dev-mentoring reviewed (changes requested):
고생많으셨습니다! 수정사항 반영해서 업데이트 부탁드립니다!

axhtl committed:
Merge remote-tracking branch 'origin/develop' into feature/order-create

axhtl committed:
feat: @RequestBody, @Validated 추가 및 의존성 추가

- OrderController에 @RequestBody, @Validated 추가
- build.gradle에 validation 의존성 추가

axhtl committed:
refactor: 축약어 변경 및 개행 정리

- Order 필드명 축약어 변경
  createdDt -> createdDate
  createdAt -> createdTime
  shippingAmt -> totalShippingPrice
  productsAmt -> totalProductsPrice
  paymentAmt -> totalPrice

- OrderService의 getOrder메서드에서 OrderResponse의 생성자 매개변수 개행 정리

- Order의 생성자 매개변수 개행 정리

axhtl committed:
refactor: Setter 어노테이션 삭제 및 RequiredArgsConstructor 어노테이션 추가

- CreateOrderRequest에서 Setter 어노테이션 삭제

- CreateOrderRequest에서 RequiredArgsConstructor 어노테이션 추가
- 필드에 final 추가

axhtl committed:
refactor: createOrderRequest를 Order 객체로 변환하여 레포지토리에 저장

- OrderService의 saveOrder 메서드에서, createOrderRequest를 Order 객체로 변환하여 레포지토리에 저장

axhtl committed:
refactor: RequiredArgsConstructor, Setter, DateTimeFormat 어노테이션 삭제

- CreateOrderRequest에서 RequiredArgsConstructor 어노테이션 삭제

- Order에서 Setter, DateTimeFormat 어노테이션 삭제

axhtl committed:
refactor: 테이블 이름 변경 orderline -> orders 및 컬럼 이름 및 타입 변경

- 컬럼 이름 변경
'created_dt' -> 'created_date'
- 컬럼 타입 변경
created_date 타입: varchar(255) -> timestamp
created_at 타입: datetime -> timestamp

axhtl committed:
refactor: 테이블 이름 변경 orderline -> orders 및 컬럼 이름 및 타입 변경

- 컬럼 이름 변경
'created_dt' -> 'created_date'
- 컬럼 타입 변경
created_date 타입: varchar(255) -> timestamp
created_at 타입: datetime -> timestamp

- Order에서 Setter, DateTimeFormat 어노테이션 삭제

axhtl committed:
Merge branch 'feature/order-create' of https://github.com/ClothingStoreService/clothstar_v1.0 into feature/order-create

axhtl committed:
refactor: createdDate 필드의 타입 변경 String->LocalDateTime 및 enum Status 추가

axhtl committed:
refactor: Order에 toOrderResponse 메서드 생성

- OrderService에서 getOrder메서드의 리턴값 수정
- OrderResponse에서 Setter 어노테이션 삭제

axhtl committed:
style: OrderResponse에 빌더 어노테이션 주석 삭제

axhtl committed:
refactor: Order에서 필드명 변경

- 필드명 변경
createdTime -> createdAt
totalShippingPrice -> shippingAmt
totalProductsPrice -> productsAmt
totalPrice -> paymentAmt

axhtl committed:
test: OrderServiceTest 추가(주문 생성만 테스트)

- 현재는 서비스에서 주문 생성만 테스트할 수 있습니다.
추가적인 로직(예: 주문 생성시 '재고 수량'이 0이라면 주문이 생성되지 않음)은 아직 구현되지 않았고, 향후 구현할 예정입니다.

axhtl committed:
refactor: Orders 테이블 컬럼명 변경 & Order 클래스 필드명 변경

axhtl committed:
docs: orderREADME.md 수정 - v1 내용까지 작성

dev-mentoring reviewed (approved):
고생하셨습니다!!
